tls/s2n_recv.c

```diff
@@ -131,6 +132,9 @@ ssize_t s2n_recv(struct s2n_connection *conn, void *buf, ssize_t size, s2n_block
             if (r == -2) {
                 conn->closed = 1;
                 *blocked = S2N_NOT_BLOCKED;
+                if (s2n_is_caching_enabled(conn->config) && conn->session_id_len) {
+                    conn->config->cache_delete(conn->config->cache_delete_data, conn->session_id, conn->session_id_len);
+                }
                 if (!bytes_read) {
                     GUARD(s2n_connection_wipe(conn));
                 }
```

Comment:
GitHub won't let me comment on the lines below this. I think we should also expire cache entries when we fail to parse records/record headers.
The catch-all line with `S2N_ERROR(S2N_ERR_IO)` is incorrect. We could be overwriting s2n_errno that was set in s2n_read_full_record(like in s2n_record_parse). I can work on a little change that fixes this.